File: src/215.数组中的第k个最大元素.rs
/*
 * @lc app=leetcode.cn id=215 lang=rust
 *
 * [215] 数组中的第K个最大元素
 */
struct Solution;
// @lc code=start
use rand::prelude::*;

#[allow(unused)]
impl Solution {
    fn find_kth_smallest(nums: &mut [i32], k: usize) -> i32 {
        let n = nums.len();
        if k == 0 {
            return *nums.into_iter().min().unwrap();
        }
        // quit sort
        let pivot = nums[0];
        let (mut i, mut j) = (1, n - 1);
        while i < j {
            while i < j && nums[i] <= pivot {
                i += 1;
            }
            while i < j && pivot <= nums[j] {
                j -= 1;
            }
            nums.swap(i, j);
        }

        // [0, mid), mid(pivot), [mid+1, n)
        let mid = if nums[i] <= pivot { i } else { i - 1 };
        // swap pivot to mid
        nums.swap(0, mid);

        // recursive
        if k == mid {
            return pivot;
        } else if k < mid {
            // sort left
            Self::find_kth_smallest(&mut nums[0..mid], k)
        } else {
            Self::find_kth_smallest(&mut nums[mid + 1..n], k - mid - 1)
        }
    }

    pub fn find_kth_largest(mut nums: Vec<i32>, k: i32) -> i32 {
        let mut rng = thread_rng();
        nums.shuffle(&mut rng);

        let n = nums.len();
        let k = n - k as usize;
        Self::find_kth_smallest(&mut nums, k)
    }
}
// @lc code=end
#[test]
fn test_solution() {
    macro_rules! test {
        ($v:tt, $k:expr, $ans:expr) => {
            assert_eq!(
                Solution::find_kth_largest(vec!$v, $k),
                $ans
            );
            eprintln!();
        }
    };
    test!([3, 2, 1], 1, 3);
    test!([3, 2, 1], 2, 2);
    test!([3, 2, 1], 3, 1);
    test!([3, 2, 1, 5, 6, 4], 2, 5);
    test!([3, 2, 3, 1, 2, 4, 5, 5, 6], 4, 4);
    test!([1], 1, 1);
}

File: src/215.数组中的第k个最大元素.py
#
# @lc app=leetcode.cn id=215 lang=python3
#
# [215] 数组中的第K个最大元素
#
from prelude import *
# @lc code=start
from typing import Tuple
import random


class Solution:
    def partition(self, nums: List[int], i: int, j: int, pivot: int) -> Tuple[int, int, int]:
        less, equal, greater = i, i, j
        # [i, less) => less
        # [less, equal) => equal
        # [greater, j) => greater
        while equal < greater:
            if nums[equal] == pivot:
                equal += 1
            elif nums[equal] < pivot:
                # swap [less], [equal]
                nums[less], nums[equal] = nums[equal], nums[less]
                less += 1
                equal += 1
            else:
                # >
                greater -= 1
                nums[equal], nums[greater] = nums[greater], nums[equal]

        # 返回 < = > 数量
        return (less - i, equal - less, j - greater)

    def findKthLargestInSlice(self, nums: List[int], i: int, j: int, k: int) -> int:
        if i + 1 == j:
            return nums[i]
        # [i, j)
        pivot = nums[random.randrange(i, j)]
        less_cnt, equal_cnt, greater_cnt = self.partition(nums, i, j, pivot)
        # 在 > 段
        if greater_cnt >= k:
            return self.findKthLargestInSlice(nums, i+less_cnt+equal_cnt, j, k)
        if greater_cnt + equal_cnt >= k:
            return pivot
        else:  # < 段
            return self.findKthLargestInSlice(nums, i, i+less_cnt, k-greater_cnt-equal_cnt)

    def findKthLargest(self, nums: List[int], k: int) -> int:
        return self.findKthLargestInSlice(nums, 0, len(nums), k)


# @lc code=end
if __name__ == '__main__':
    def test(input, k, expected):
        calc = Solution().findKthLargest(input[::], k)
        if calc != expected:
            print(f'case failed: `{input}`')
            print(f'  calc     = `{calc}`')
            print(f'  expected = `{expected}`')
            exit(1)
    test([3, 2, 1, 5, 6, 4], 2, 5)
    test([3, 2, 3, 1, 2, 4, 5, 5, 6], 4, 4)
